### ISSUE TITLE
chromium-ozone-wayland: eglCreateContext: Requested version is not supported

### ISSUE THREAD

**Issue by MastaG** [OSSystems/meta-browser#849]
When I run chromium-ozone-wayland on my armv7 board using the Panfrost mesa drivers, I always get the following error:
```
ERROR:gl_display.cc(497)] EGL Driver message (Error) eglCreateContext: Requested version is not supported
```
The browser runs fine otherwise and seems fully accelerated except for WebGL.
Is it possible I could get more verbose information regarding the above error?
My chromium flags:
`--ozone-platform=wayland --enable-wayland-ime --use-gl=angle --use-angle=gles-egl --use-cmd-decoder=passthrough --ignore-gpu-blocklist --ignore-gpu-blacklist --enable-accelerated-video-decode --enable-features=VaapiVideoDecoder,VaapiVideoEncoder --gtk-version=4 --disable-gpu-memory-buffer-video-frames`

By the way: `--use-gl=egl` doesn't seem to work anymore and needs to be replaced with: `--use-gl=angle --use-angle=gles-egl` or `--use-gl=angle --use-angle=gl-egl`

**Comment by MastaG** [comment]
With `--use-angle=gl-egl` I'm getting the same EGL error.
Is it possible to compile Chromium with legacy EGL support (`--use-gl=egl`)?
Because Angle doesn't seem to work for my platform.

**Comment by msisov** [comment]
> With `--use-angle=gl-egl` I'm getting the same EGL error. Is it possible to compile Chromium with legacy EGL support (`--use-gl=egl`)? Because Angle doesn't seem to work for my platform.

Can you try https://github.com/OSSystems/meta-browser/pull/850 ?

**Comment by MastaG** [comment]
Yes I can only run with Angle now. So I only have `--use-angle=gles-egl`.
When I use `--use-gl=egl`  I'm getting a message that it isn't supported any longer.

Hardware acceleration works for the most part when using Angle.
It's just that it will display the: `ERROR:gl_display.cc(497)] EGL Driver message (Error) eglCreateContext: Requested version is not supported` twice.
And WebGL seems to only thing that fails to work.

See the report:

[about-gpu-2024-11-19T11-15-51-085Z.txt](https://github.com/user-attachments/files/17814344/about-gpu-2024-11-19T11-15-51-085Z.txt)

**Comment by msisov** [comment]
Can you try `--use-angle=gl-egl` ?

**Comment by MastaG** [comment]
Yeah that's probably it.
I now reckon that the Mesa Panfrost driver for me Mali GPU, only supports OpenGLES 2.0 and OpenGL 2.1.

I've switched to the proprietary binary driver from ARM which unfortunately only supports OpenGLES 3.1 on X11 and WebGL is working just fine when running on the Ozone X11 platform.

So the reason for the `ERROR:gl_display.cc(497)] EGL Driver message (Error) eglCreateContext: Requested version is not supported` is probably OpenGL/ES 3.x or higher missing from the Mesa driver.

So I can confirm that the change in PR https://github.com/OSSystems/meta-browser/pull/850 works across both X11 and Wayland.
It's just that my platform is missing OpenGL/ES 3.x on Wayland being a requirement for WebGL.

**Comment by MaxIhlenfeldt** [comment]
Does https://webglsamples.org/aquarium/aquarium.html work for you? afaict that's using WebGL 1, and that should only require OpenGL ES 2.0 [according to MDN](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext).

**Comment by MastaG** [comment]
> Does https://webglsamples.org/aquarium/aquarium.html work for you? afaict that's using WebGL 1, and that should only require OpenGL ES 2.0 [according to MDN](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext).

That was the exactly the demo I was trying.
However when I run chromium using the Mesa Panfrost driver (GL2.1 ES2.0), the "`ERROR:gl_display.cc(497)] EGL Driver message (Error) eglCreateContext: Requested version is not supported`" will print immediately on the terminal when starting the browser.
I've tried both `--use-angle=gles-egl` and `--use-angle=gl-egl`.
This probably means that Angle isn't happy about it.

Then when I open the aquarium demo, the webpage will display:
`Status: Could not create a WebGL context, VENDOR = 0xffff, DEVICE = 0xffff, GL_VENDOR = Google Inc. (Mesa), GL_RENDERER = ANGLE (Mesa, Mali-T620 (Panfrost), OpenGL ES 2.0 Mesa 24.2.3-1ubuntu1), GL_VERSION = 24.2.3, Sandboxed = no, Optimus = no, AMD switchable = no, Reset notification strategy = 0x0000, ErrorMessage = BindToCurrentSequence failed: .`
While the terminal will say:
`ERROR:context_group.cc(179)] ContextResult::kFatalFailure: ES3 is blocklisted/disabled/unsupported by driver.`

That's why I think Angle will fail to initialize the WebGL backend, because I don't have OpenGL(ES) 3.x or higher.

**Comment by WujianSun** [comment]
Hi @msisov ,
Do you any differences bewteen --use-angle=gles-egl and --use-angle=gles?

**Comment by TheNathan0** [comment]
@MastaG 

As one point I had working WebGL with ANGLE on Wayland.  The GPU being a Mali T628 using Mesa Panfrost and the old OpenGL and OpenGL ES APIs of course.  If I recall correctly (don't quote me if I'm incorrect here) it worked and it laos using ANGLE (I did not specify any kind of any OpenGL/OpenGL ES API flags).  I recall still getting `ERROR:gl_display.cc(497)] EGL Driver message (Error) eglCreateContext: Requested version is not supported` though.  Didn't need to disable sandboxing either.

However I did encounter serious and severe visual artifacting, bugs, etc when using '--use-gl=egl' ([which required a patched version of Wayland](https://aur.archlinux.org/packages/wayland-chromium)) back when the flag still existed.  I think this occurred with anything WebGL related under such a setup though.

Not a Chromium expert, but maybe some patches could bring back WebGL on Wayland + Mesa Panfrost + Mali T628.  Not sure if WebGL no longer working on legacy OpenGL/OpenGL ES was intentional or not.

>  I now reckon that the Mesa Panfrost driver for me Mali GPU, only supports OpenGLES 2.0 and OpenGL 2.1.
Concerning this, it's due to the fact that all [Midgard v4 GPUs](https://docs.mesa3d.org/drivers/panfrost.html) (including the Mali T620 series) that had OpenGL ES 3.1 support with the ARM driver did not have hardware Multiple Render Targets (MRT) capabilities.  Instead, the ARM proprietary driver emulated it, something that no one is currently interested in supporting with the Panfrost driver.  Midgard v5 GPUs and newer all have hardware MRT support.
